extensions/Worker.Extensions.ServiceBus/src/ServiceBusReceivedMessageConverter.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using Azure.Messaging.ServiceBus;
+using Azure.Messaging.ServiceBus.Primitives;
+using Microsoft.Azure.Functions.Worker.Converters;
+using Microsoft.Azure.Functions.Worker.Core;
+
+namespace Microsoft.Azure.Functions.Worker;
+
+internal class ServiceBusReceivedMessageConverter : IInputConverter
+{
+    public ValueTask<ConversionResult> ConvertAsync(ConverterContext context)
+    {
+        if (context is null)
+        {
+            return new ValueTask<ConversionResult>(ConversionResult.Unhandled());
+        }
+
+        return context.Source switch
+        {
+            ModelBindingData binding => ConvertToServiceBusReceivedMessage(context, binding),
+            _ => new ValueTask<ConversionResult>(ConversionResult.Unhandled()),
+        };
+    }
+
+    private ValueTask<ConversionResult> ConvertToServiceBusReceivedMessage(ConverterContext context, ModelBindingData binding)
+    {
+        // The lock token is a 16 byte GUID
+        const int lockTokenLength = 16;
+
+        ReadOnlyMemory<byte> bytes = binding.Content.ToMemory();
```

Comment:
I'd like us to follow a similar pattern to the blob converter here for the content-type. We know that it's currently a stream based on the changes in your PR, but I would prefer we future proof this a bit:

Here's what we do in the blob converter today:

```csharp
       private Dictionary<string, string> GetBindingDataContent(ModelBindingData bindingData)
        {
            return bindingData?.ContentType switch
            {
                Constants.JsonContentType => new Dictionary<string, string>(bindingData?.Content?.ToObjectFromJson<Dictionary<string, string>>(), StringComparer.OrdinalIgnoreCase),
liliankasem marked this conversation as resolved.
                _ => throw new InvalidOperationException($"Unexpected content-type. Currently only {Constants.JsonContentType} is supported.")
            };
        }
```

Reply:
This doesn't really apply for Service Bus as the content must be opaque bytes.

Reply:
That's fair, but something feels off about ignoring the content-type  😆

Reply:
Hah fair point.